### DISCUSSION TITLE
✨ Improve module debugging experience

### DISCUSSION THREAD

**Discussion by sagikazarmark** [dagger/dagger#12142] [Ideas]
### What are you trying to do?

With the introduction of modules, developers can now reuse pieces of code for building their workflows, either built by them or other developers.

I mostly used my own modules so far, so adding new features or debugging module functionality wasn't particularly difficult.

When it comes to tinkering with third-party modules, especially for someone who isn't very familiar with how modules work, this may be more difficult.

This is how you do it today:

- Look into `dagger.json`
- Find the repository for the dependency
- Clone the repository
- Copy the relevant subdirectory (or the whole repository) into your root/contextual module somewhere
- Replace the reference to the module with the local path

There are multiple issues with this:

- It's a tedious process
- Currently, you lose the original reference to the module (you need to look it up from Git if you want to go back to the original source)

It would be nice if the CLI could help with speeding up the process:

- `dagger vendor` or `dagger replace` command to download a particular dependency
- a new `replace` (or similar) field in `dagger.json` so the original reference is preserved (similar to how `go.mod` works)

This would make it easier for people to check out a module, look at the source code and live debug it to figure out any issues. (In the future, making contributions easier could become a goal, but for the short term I'd say it's out of scope)

Original message on Discord: https://discord.com/channels/707636530424053791/1120503349599543376/1303760702892150844

**Comment by mihaelmiklec** [comment]
That would be quite handy! I’d like to add that `vendor` would probably be the best name for this command, as it's a commonly used term for including third-party dependencies directly in the codebase. Mimicking the `replace` directive from `go.mod` would be familiar to Go developers, but it would need to be adjusted for the JSON format.

The vendor command could pull in specified dependencies and automatically add a `replacement` field in `dagger.json` for each dependency, like so:
```json
"dependencies": [
  {
    "name": "some-mod",
    "source": "github.com/org/repo/some-mod@1eed56a12ffcd897efe2dls5as7d4dd0c015d29f89d9",
    "replacement": "./vendored/org/repo/some-mod"
  }
]

```
